### PR TITLE
Adding support for creating ASGs with multiple instance types, instance diversification via API

### PR DESCRIPTION
**Pull request by pdk27**
## Summary
Adding support for creating ASGs with multiple instance types, instance diversification via API
   - support for creating ASG with mixed instances policy (MIP)
   - update caching to track launch templates created with MIP
   - parse MIP data from AWS API and translate to Clouddriver API response
   - track only single instance type case with MIP and multiple types tracked in nested MIP JSON value for backwards compatibility
   - support for multiple instance type validation / processing / deriving defaults
   - add validations and tests (added tests with detailed data - launch template / MIP to also serve as samples) 
   - rename spotPrice to spotMaxPrice for clarity, in non-customer exposed places
   
## Issue
https://github.com/spinnaker/spinnaker/issues/5989

## Related AWS docs:
[Launching ASG with MIP and it's options](https://docs.aws.amazon.com/autoscaling/ec2/userguide/asg-purchase-options.html)
[MIP API Ref](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_MixedInstancesPolicy.html)

## Implementation details and testing :
<details>
  <summary>Create ASG with MIP. What new fields in API request looks like (Click to Expand)</summary>

image image
</details>

<details>
  <summary>Extract ASG launch template and related information from MIP. API response snippets (Click to Expand)</summary>

```
{
  "asg": {
    "autoScalingGroupName": "hello-mip-v002",
    "maxSize": 5,
    "minSize": 3,
    "mixedInstancesPolicy": {
       "instancesDistribution": {
         "onDemandAllocationStrategy": "prioritized",
         "onDemandBaseCapacity": 1,
         "onDemandPercentageAboveBaseCapacity": 50,
         "spotAllocationStrategy": "capacity-optimized",
         "spotMaxPrice": "1"
       },
       "launchTemplate": {
          "launchTemplateSpecification": {
             "launchTemplateId": "lt-123",
             "launchTemplateName": "hello-mip-v002-123",
             "version": "$Latest"
          },
          "overrides": [
          {
             "instanceType": "t2.medium",
             "weightedCapacity": "2"
          },
          {
             "instanceType": "t2.large",
             "weightedCapacity": "4"
          }
          ]
       }
    },
    ...
  },
  ...
  "mixedInstancesPolicy": {
     "allowedInstanceTypes": [
         "t2.large",  "t2.medium"
     ],
     "instancesDiversification":{
        "onDemandAllocationStrategy":"prioritized",
        "onDemandBaseCapacity":1,
        "onDemandPercentageAboveBaseCapacity":50,
        "spotAllocationStrategy":"capacity-optimized",
        "spotMaxPrice":"1"
     },
     "launchTemplateOverridesForInstanceType":[
           {
              "instanceType":"t2.medium",
              "weightedCapacity":"2"
           },
           {
              "instanceType":"t2.large",
              "weightedCapacity":"4"
           }
        ],
     "launchTemplates": [
      {
        "createTime":1614276214000,
        "createdBy":"arn:aws:sts::0123456789:assumed-role/SpinnakerManaged/Spinnaker",
        "defaultVersion":true,
        "launchTemplateData":{
          ...
           "creditSpecification":{
              "cpuCredits":"standard"
           },
           "instanceType":"t2.small",
           "networkInterfaces":[
              {
                 "deviceIndex":0,
                 "groups":["sg-123" ],
                 "ipv6AddressCount":0,
                 "ipv6Addresses":[ ],
                 "privateIpAddresses":[ ]
              }
           ],
           ...
        },
        "launchTemplateId":"lt-123",
        "launchTemplateName":"hello-mip-v002-123",
        "versionNumber":1
      }
     ]
  }
  ...
}
```
</details>

* Caching of Launch templates created by MIP and adding appropriate relationships

* Validations: 
    > `spotInstancePools` supported only with `spotAllocationStrategy` 'lowest-price':
   
```
"responseBody": "{\"error\":\"Bad Request\",\"errors\":[\"basicAmazonDeployDescription.spotInstancePools.not.supported.for.spotAllocationStrategy\"],\"message\":\"Validation Failed\",\"status\":400,\"timestamp\":1614708699422}",
"kind": "HTTP",
"error": "Bad Request",
"errors": [
     "basicAmazonDeployDescription.spotInstancePools.not.supported.for.spotAllocationStrategy"
],
"url": "http://localhost:7002/aws/ops?clientRequestId=XXX",
"status": 400
```

  > if `unlimitedCpuCredits` is specified, all instance types are expected to support bursting, in order to ensure compatibility among config parameters:
   
```
"responseBody": "{\"error\":\"Bad Request\",\"errors\":[\"basicAmazonDeployDescription.bursting.not.supported.by.instanceType\"],\"message\":\"Validation Failed\",\"status\":400,\"timestamp\":1617370445152}",
"kind": "HTTP",
"error": "Bad Request",
"errors": [
"basicAmazonDeployDescription.bursting.not.supported.by.instanceType"
],
"url": "http://localhost:7002/aws/ops?clientRequestId=3523ae7dd7176cfb567210229d980bd5fbf7dcbf99a72ab44232dad6ed48dd4c",
"status": 400
```

* View ASG with MIP in Deck (PR: https://github.com/spinnaker/deck/pull/8960)

### CODE REVIEW

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonServerGroup.groovy [hunk 1]
```diff
@@ -37,6 +38,15 @@ class AmazonServerGroup implements ServerGroup, Serializable {
   Map<String, Object> image
   Map<String, Object> launchConfig
   Map<String, Object> launchTemplate
+
+  /**
+   * The fields marked @Alpha are subject to change in the future due to addition of newer ASG features like multiple launch templates.
+   */
+  @Alpha
```
Comment (pdk27) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonServerGroup.groovy:45]: marked as `Alpha` as the data model will probably change in the future when multiple launch templates support is added.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonServerGroup.groovy [hunk 2]
```diff
@@ -132,9 +147,27 @@ class AmazonServerGroup implements ServerGroup, Serializable {
     return securityGroups
   }
 
+  /**
+   * Get a single instance type for a server group.
+   * Used for the case of single instance type i.e. ASG with launch config, launch template, mixed instances policy without overrides
+   *
+   * For the case of multiple instance types i.e. ASG MixedInstancesPolicy with overrides, gather types from mixedInstancesPolicy.
```
Comment (pdk27) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonServerGroup.groovy:154]: Would love to hear if you can think of a better way to handle this inconsistency.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonServerGroup.groovy [hunk 3]
```diff
@@ -37,6 +38,15 @@ class AmazonServerGroup implements ServerGroup, Serializable {
...
   Map<String, Object> launchConfig
   Map<String, Object> launchTemplate
+
+  /**
+   * The fields marked @Alpha are subject to change in the future due to addition of newer ASG features like multiple launch templates.
+   */
+  @Alpha
+  Map<String, Object> mixedInstancesPolicy
+  @Alpha
+  Map<String, Object> launchTemplateSpecification
```
Comment (pdk27) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonServerGroup.groovy:48]: There will probably be more launch settings that will need to be extracted out of either `launchTemplate` or `mixedInstancesPolicy` going forward. Hence the `Alpha` annotation.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ClusterCachingAgent.groovy [hunk 4]
```diff
@@ -742,8 +742,8 @@ class ClusterCachingAgent implements CachingAgent, OnDemandAgent, AccountAware,
     final String cluster
     final String serverGroup
     final String vpcId
-    final String launchConfig
-    final String launchTemplate
+    final String launchConfigKey
+    final String launchTemplateKey
```
[the remark this answers is not in the document]
Comment (pdk27) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ClusterCachingAgent.groovy:746]: oops. My bad! I meant to remove these changes. Cleaning it up.

**Review by robzienert — APPROVED** [review]

**Pull request merged**